extensions/amp-a4a/0.1/test/test-amp-a4a.js

```diff
@@ -124,6 +139,198 @@ describe('amp-a4a', () => {
+            a4a, 'extractCreativeAndSignature');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(extractCreativeAndSignatureSpy.calledOnce,
+		 'extractCreativeAndSignatureSpy called exactly once').to.be.true;
+        });
+      });
+    });
+
+    it('should throw error if extractCreativeAndSignature not implemented', () => {
+      viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).onFirstCall().returns(Promise.resolve(mockResponse));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+        const a4a = new MockA4AImplNoExtractCreativeAndSignature(a4aElement);
+	const extractCreativeAndSignatureSpy = sandbox.spy(
+            a4a, 'extractCreativeAndSignature');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(extractCreativeAndSignatureSpy.calledOnce,
+		 'extractCreativeAndSignatureSpy called exactly once').to.be.true;
+	  expect(extractCreativeAndSignatureSpy.exceptions.length).to.equal(1);
+	});
+      });
+    });
+  });
+
+  describe('#getAdUrl', () => {
+    it('should be called once', () => {
+     viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).onFirstCall().returns(Promise.resolve(mockResponse));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+        const a4a = new MockA4AImpl(a4aElement);
+	const getAdUrlSpy = sandbox.spy(
+            a4a, 'getAdUrl');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
```

Comment:
Yeah...  This is an instance of what I'd like to get *away* from, in general: reaching into the guts of the implementation and picking up specific fields.  If we significantly refactor the promise chain (as we're hoping to), this could conceivably go away.

It's fine for some individual tests, but I'd _really_ like us to get up to being able to do end-to-end tests where we just insert the `amp-ad` element on the page, let the AMP runtime do its stuff to it, and then check the resulting DOM.  (And manipulate various failures by manipulating the systems that A4A interacts with -- like XHR or the "ad network-provided" code -- that aren't actually part of A4A itself.)  I have a question out to Dima on how to do so, so for the moment this style is probably the best we can do.